### PR TITLE
Implement search API (fixes #929)

### LINKED ISSUE #929: Search is not implemented in Twitter v.1 - compatible API
I encounted this problem testing Friendica work with [AndStatus Android client](https://github.com/andstatus/andstatus/wiki).
E.g. valid Search request URL="https://pirati.ca/api/search.json?count=200&q=AndStatus" returns error 404 page.
See corresponding issue in AndStatus: [support for friendica](https://github.com/andstatus/andstatus/issues/55)

### PR DESCRIPTION
**Pull request by Rudloff**
This PR adds two API endpoints:
* `search` (from the [GNU Social API](https://gnusocial.net/doc/twitterapi))
* `search/tweets` (from the [Twitter API](https://developer.twitter.com/en/docs/tweets/search/api-reference/get-search-tweets))
As the parameters are very similar, they both use the same function.
All the GNU Social parameters are implemented but several Twitter parameters are missing.

I'm not very familiar with the Friendica database and most of the SQL code is borrowed from other functions so please tell me if I missed something.

### CODE REVIEW

File: include/api.php [hunk 1]
```diff
@@ -1485,6 +1485,71 @@ function api_users_search($type)
...
+ *
+ * @return array|string
+ * @throws UnauthorizedException
+ * @throws BadRequestException
+ */
+function api_search($type)
+{
+	$data = array();
+
+	if (x($_REQUEST, 'q')) {
```
Comment (annando) [inline include/api.php]: You could throw the exeception if "q" was not set at the beginning. So you don't need to indent the rest.

File: include/api.php [hunk 2]
```diff
@@ -1485,6 +1485,71 @@ function api_users_search($type)
...
+		}
+
+		$r = q(
+			"SELECT %s
+			FROM `item` %s
+			WHERE %s AND (`item`.`uid` = 0 OR (`item`.`uid` = %s AND NOT `item`.`global`))
+			AND `item`.`body` REGEXP '%s'
+			$sql_extra
+			AND `item`.`id`>%d
+			GROUP BY `item`.`uri`, `item`.`id`
```
Comment (annando) [inline include/api.php]: grouping by item.id doesn't seem to be a good idea - this field is unique, so grouping after it doesn't make sense. Additionally I think that you shouldn't need the grouping at all here.

File: include/api.php [hunk 3]
```diff
@@ -1485,6 +1485,71 @@ function api_users_search($type)
...
+			ORDER BY `item`.`id` DESC LIMIT %d ,%d ",
+			item_fieldlists(),
+			item_joins(),
+			item_condition(),
+			intval(local_user()),
```
Comment (annando) [inline include/api.php]: We have "api_user" as a function for the currently active user in API calls.

File: include/api.php [hunk 4]
```diff
@@ -1485,6 +1485,71 @@ function api_users_search($type)
...
+			dbesc(protect_sprintf(preg_quote($_REQUEST['q']))),
```
Comment (annando) [inline include/api.php]: What is the sense of ```protect_sprintf``` here?

Comment (Rudloff) [reply include/api.php]: This part was borrowed from here: https://github.com/friendica/friendica/blob/develop/mod/search.php#L208
But you're right, it does not seem useful here.

File: include/api.php (hunk 2 above)
Comment (MrPetovan) [reply include/api.php]: Grouping by a unique field can make sense when you’re doing a one-to-many join but only need a single row per unique value, usually when you need the result of an aggregate function like COUNT or MAX.

Comment (annando) [reply include/api.php]: Yeah, but this isn't the case here.

File: include/api.php (hunk 4 above)
Comment (annando) [reply include/api.php]: I just looked at that the function is doing - it seems to make sense. But it really looks weird.

File: include/api.php (hunk 2 above)
Comment (MrPetovan) [reply include/api.php]: On mobile, I can’t see the code past your comment, so I couldn’t see the SELECT fields! 😅

**Comment by Rudloff** [comment]
I updated the PR with your suggestions.

(One issue I noticed with `dba::p()` is that parameters are always treated as strings, so I could not use it to insert the `LIMIT` values.)

**Comment by MrPetovan** [comment]
Yeah, it’s a known limitation. I tried to implement parameter type in bindParam but it broke a lot of requests, so you still need to use concatenation and intval for now.

File: include/api.php [hunk 5]
```diff
@@ -1500,49 +1500,48 @@ function api_search($type)
...
+		$sql_extra
+		AND `item`.`id`>?
+		ORDER BY `item`.`id` DESC LIMIT ".intval($start)." ,".intval($count)." ",
+		intval(api_user()),
+		$_REQUEST['q'],
+		intval($since_id)
+	);
+
+	$statuses = array();
+	while ($row = dba::fetch($r)) {
```
Comment (annando) [inline include/api.php]: This could be made easier by using ```dba::inArray($r);```

**Comment by annando** [comment]
A way to avoid the limitations of not being able to use prepared values to "limit" would be dropping support for MySQLi and only using PDO. (AFAIK this is able doing so)

File: include/api.php (hunk 5 above)
Comment (Rudloff) [reply include/api.php]: Thanks! I was looking for a function doing this but did not find it (I was expecting it to be called something like `fetchAll`).

File: include/api.php [hunk 6]
```diff
@@ -1485,6 +1485,65 @@ function api_users_search($type)
...
+
+	$r = dba::p(
+		"SELECT ".item_fieldlists()."
+		FROM `item` ".item_joins()."
+		WHERE ".item_condition()." AND (`item`.`uid` = 0 OR (`item`.`uid` = ? AND NOT `item`.`global`))
+		AND `item`.`body` REGEXP ?
+		$sql_extra
+		AND `item`.`id`>?
+		ORDER BY `item`.`id` DESC LIMIT ".intval($start)." ,".intval($count)." ",
+		intval(api_user()),
```
Comment (annando) [inline include/api.php]: you don't need ```intval``` here - the new dba functions are doing this automatically.

File: include/api.php [hunk 7]
```diff
@@ -1485,6 +1485,65 @@ function api_users_search($type)
...
+		$_REQUEST['q'],
+		intval($since_id)
```
Comment (annando) [inline include/api.php]: Like above.

**Comment by annando** [comment]
@MrPetovan You are okay with it?

**Comment by Rudloff** [comment]
@annando Yes, the regexp still works correctly.
But I just noticed it allows to use regex directly in the query (e.g. `api/search.json?q=^foo`) which is not something the Twitter API supports. (Twitter uses [search operators](https://developer.twitter.com/en/docs/tweets/search/guides/standard-operators) instead but this looks like a lot of work to implement.)

Should we use something more restrictive like this?
```diff
diff --git a/include/api.php b/include/api.php
index 39a1f0318..c359c65f1 100644
--- a/include/api.php
+++ b/include/api.php
@@ -1526,7 +1526,7 @@ function api_search($type)
                "SELECT ".item_fieldlists()."
                FROM `item` ".item_joins()."
                WHERE ".item_condition()." AND (`item`.`uid` = 0 OR (`item`.`uid` = ? AND NOT `item`.`global`))
-               AND `item`.`body` REGEXP ?
+               AND `item`.`body` LIKE CONCAT('%',?,'%')
                $sql_extra
                AND `item`.`id`>?
                ORDER BY `item`.`id` DESC LIMIT ".intval($start)." ,".intval($count)." ",
```

**Comment by MrPetovan** [comment]
Ideally we should use a FULLTEXT search, InnoDB and MariaDB both support fulltext indices. Both the REGEXP and the LIKE will fail to return results for individual terms separated by a space.

Until we use a fulltext index on body, we should use the fastest comparison, I believe LIKE.

**Comment by annando** [comment]
I guess that a ```LIKE``` uses less CPU performance than a ```REGEXP```. So just do it.

**Comment by annando** [comment]
```FULLTEXT``` does have huge problems with the way that our database is constructed. You cannot combine ```FULLTEXT``` and other indexes. So the system will always have to do a table scan.

**Pull request merged**